### ISSUE TITLE
Text rendering overlap on scroll in a terminal editor

### ISSUE THREAD

**Issue by pariola** [zeit/hyper#3391]
- [x] I am on the [latest](https://github.com/zeit/hyper/releases/latest) Hyper.app version

- **OS version and name**: Ubuntu 18.04.1 LTS 
- **Hyper.app version**: 2.1.0

## Issue

I noticed if i use a terminal editor like nano, when i scroll down and happen to scroll back up the texts overlap

**Before Scroll**
![before](https://user-images.githubusercontent.com/11618490/50730077-0f61e400-1145-11e9-8d89-0340ebc928c3.png)

**After Scroll**
![after](https://user-images.githubusercontent.com/11618490/50730088-4932ea80-1145-11e9-837e-0b741b65b4f5.png)

**Comment by j-f1** [comment]
As a workaround, you can hit `^L` to tell Nano to redraw the entire display.

**Comment by stu177** [comment]
Getting the same on Ubuntu 18.04.1 and Hyper 3.0.0-canary.4

The same behaviour occurs in htop when data is refreshing.

**Comment by connorbrez** [comment]
Also can reproduce this

**Comment by meandmymind** [comment]
Same issue.
Hyper 2.1.2 (stable) with no extensions
Ubuntu 18.04 + KDE

**Comment by ano95** [comment]
Can confirm. Linux Mint 19, hyper 2.1.2.

**Comment by voquangphu** [comment]
Same here, happens every time I use nano.
MacOS Mojave, version 2.1.2 (stable).

**Comment by b3njaminall3n** [comment]
Still an issue, happens when paging up/down in nano. Repainting with Ctrl+L fixes the screen until you scroll again, as @j-f1 mentions.
Hyper.app 3.0.0, no plugins, macOS 10.14.4

**Comment by PandawanFr** [comment]
Happens for me too on macOS 10.14.5 with Hyper v3.0.2 (stable).

Many resources online said that you should set the `TERM` env variable. The default for Terminal.app is `xterm-256color` which works fine when using it. But when I set it in Hyper.app (both through the `.hyper.js` config and as an export in the `.bashrc`), it still breaks. 

I've seen multiple issues on GitHub that mention this including #3629, #3491, #3142, #3664, #3596. If there are that many people mentioning this issue, I think that it's a pretty important bug. 

Note: This is not a bug specific to nano since other issues (#3629) seem to be mentioning vim.

**Comment by stu177** [comment]
Yup would be nice if this is addressed. Had this issue for 6 months now and it can be incredibly frustrating.

**Comment by stu177** [comment]
For me, it seems if you set TERM to screen the issue does not occur.

For example, the issue occurs with:

`TERM=xterm nano ~/.bashrc`

But not with:

`TERM=screen nano ~/.bashrc`

Anyone else able to check if this does the same for them?

Possibly narrows the issue down a bit.

**Comment by abdilahrf** [comment]
its happen with me , no idea how to fix this!

**Comment by mnpenner** [comment]
```
TERM=screen nano foo
```

Still leaves garbage on the screen: https://i.imgur.com/gLfLkRZ.png

**Comment by TulioAbreu** [comment]
Im having the same problem on Ubuntu 18.04.3 LTS, nano 2.9.3 and Hyper 3.0.2 (stable).

**Comment by MarlBurroW** [comment]
Workaround for macos + zsh users:
Add `export TERM=screen` in `~/.zshrc`

**Comment by stu177** [comment]
> ```
> TERM=screen nano foo
> ```
> 
> Still leaves garbage on the screen: https://i.imgur.com/gLfLkRZ.png

I think this could be a different issue with the screen not clearing on change rather than specifically with scrolling. Are you able to get any issues to occur when scrolling with `TERM` set to `screen`?

**Comment by mnpenner** [comment]
@stu177 Oh, nevermind then. I don't think I've noticed that specific issue. Is there another ticket for what I'm describing? It doesn't happen in bash.exe https://i.imgur.com/Dd5owMn.png Perhaps https://github.com/zeit/hyper/issues/3792

**Comment by majordookie** [comment]
Same issue here

Opening file in nano

![image](https://user-images.githubusercontent.com/7274185/66689858-cbaf8400-ec5a-11e9-9a5f-194597017b6f.png)

Paging down all the way to bottom

![image](https://user-images.githubusercontent.com/7274185/66689867-d4a05580-ec5a-11e9-8077-41c709e549dc.png)

Exiting nano 

![image](https://user-images.githubusercontent.com/7274185/66689890-f4d01480-ec5a-11e9-9887-9056b4d897d2.png)

**Comment by majordookie** [comment]
Same issue here

Opening file in nano (no issues)

![image](https://user-images.githubusercontent.com/7274185/66689909-1f21d200-ec5b-11e9-8574-41bc38eff656.png)

Paging down to bottom, artifacts left behind 

![image](https://user-images.githubusercontent.com/7274185/66689940-3b257380-ec5b-11e9-9d04-cc18024ccd88.png)

Exiting nano, artifacts are still there

![image](https://user-images.githubusercontent.com/7274185/66689957-51cbca80-ec5b-11e9-9aeb-77552dd29962.png)

**Comment by Stanzilla** [comment]
Can you try with a CI build of current canary or maybe compile one yourself?

**Comment by majordookie** [comment]
> Can you try with a CI build of current canary or maybe compile one yourself? Or setting `webGLRenderer: false,` in the config.

Tried with canary 3.0.2 and webGL off, the issue still occurs.

**Comment by Stanzilla** [comment]
No, I meant a CI build of the canary branch

**Comment by rwsmith** [comment]
Thanks for help @Stanzilla , how do I find the CI build? I also have this issue and would like to help.

**Comment by Stanzilla** [comment]
Windows https://ci.appveyor.com/project/zeit/hyper/builds/28383913/artifacts
macOS: https://app.circleci.com/jobs/github/zeit/hyper/2831/artifacts

**Comment by rwsmith** [comment]
Thanks, this fixed it!

**Comment by LabhanshAgrawal** [comment]
From the last comment it seems like this is fixed, maybe it was an issue with the old xterm?
A new [canary](https://github.com/zeit/hyper/releases/tag/v3.1.0-canary.4) for Mac(use dmg) and Linux was released (Windows users can try it out from [here](https://ci.appveyor.com/project/chabou/hyper-1b303/builds/30940651/artifacts))
Can you all please give it a try and confirm if this issue still there.

**Comment by MarlBurroW** [comment]
Fixed for me
Tested on 3.1.0-canary.4

**Comment by LabhanshAgrawal** [comment]
> Fixed for me 👍
> Tested on 3.1.0-canary.4 (MacOSX 10.13.6)

Without the `export TERM=screen` workaround?

**Comment by MarlBurroW** [comment]
No, I removed the workaround before my test.
Works with TERM=xterm-256color

**Comment by zaneclaes** [comment]
@Stanzilla perhaps I'm confused, but the official version is still at `3.0.2` some 6 months after the build... and the build page to which you linked no longer has the official download, so I'm unsure how to acquire v`3.1.0`...

**Comment by Stanzilla** [comment]
Can you try without any plugins installed?

**Comment by zaneclaes** [comment]
Commented out all plugins from the config and restarted. No change, except the theme looks different.
<img width="710" alt="Screen Shot 2020-05-04 at 3 20 04 PM" src="https://user-images.githubusercontent.com/2084821/81014824-cf833d80-8e1a-11ea-8191-81a463837a3a.png">
Downgraded back to 3.0.2 and everything worked fine. Changed the `TERM=screen` and the original bug went away.

**Comment by Stanzilla** [comment]
Interesting, do you have webGL enabled or disabled? Should be an option in the file as well if you don't have it, it's off.

**Comment by zaneclaes** [comment]
Disabled, AFAICT. Here's my full config with superfluous comments removed & plugins re-enabled:
```
module.exports = {
  config: {
    updateChannel: 'stable',
    fontSize: 18,
    fontFamily: '"Inconsolata Nerd Font", Menlo, "DejaVu Sans Mono", Consolas, "Lucida Console", monospace',
    fontWeight: 'normal',
    fontWeightBold: 'bold',
    lineHeight: 1,
    letterSpacing: 0,
    cursorColor: 'rgba(248,28,229,0.8)',
    cursorAccentColor: '#000',
    cursorShape: 'BLOCK',
    cursorBlink: false,
    foregroundColor: '#fff',
    backgroundColor: '#000',
    selectionColor: 'rgba(248,28,229,0.3)',
    borderColor: '#333',
    css: '',
    termCSS: '',
    showHamburgerMenu: '',
    showWindowControls: '',
    padding: '12px 14px',
    colors: {
      black: '#000000',
      red: '#C51E14',
      green: '#1DC121',
      yellow: '#C7C329',
      blue: '#0A2FC4',
      magenta: '#C839C5',
      cyan: '#20C5C6',
      white: '#C7C7C7',
      lightBlack: '#686868',
      lightRed: '#FD6F6B',
      lightGreen: '#67F86F',
      lightYellow: '#FFFA72',
      lightBlue: '#6A76FB',
      lightMagenta: '#FD7CFC',
      lightCyan: '#68FDFE',
      lightWhite: '#FFFFFF',
    },

    shell: '',
    shellArgs: ['--login'],
    env: {},
    bell: 'SOUND',

    visor: {
      hotkey: 'Cmd+\\',
      position: 'right', // or left, right, bottom
    },

    paneNavigation: {
      debug: false,
      hotkeys: {
        navigation: {
          up: 'cmd+shift+up',
          down: 'cmd+shift+down',
          left: 'cmd+shift+left',
          right: 'cmd+shift+right'
        },
        jump_prefix: 'cmd+alt', // completed with 1-9 digits
        permutation_modifier: 'shift', // Added to jump and navigation hotkeys for pane permutation
        maximize: 'meta+enter'
      },
      showIndicators: true, // Show pane number
      indicatorPrefix: '^⌥', // Will be completed with pane number
      indicatorStyle: { // Added to indicator <div>
        position: 'absolute',
        top: 0,
        left: 0,
        fontSize: '10px'
      },
      focusOnMouseHover: false,
      inactivePaneOpacity: 0.3 // Set to 1 to disable inactive panes dimming
    },
  },

  opacity: 0.85,

  plugins: [
    "hyper-pane",
    "hyper-materialshell",
    "hyperterm-visor"
  ],

  localPlugins: [],

  // OSX SPECIFIC
  modifierKeys: {
    cmdIsMeta: false, // default
    altIsMeta: true // default
  },

  // DVORAK Keymap
  keymaps: {
    'tab:new': 'cmd+t',
    'tab:close': 'cmd+shift+w',
    'tab:next': 'cmd+e',
    'tab:prev': 'cmd+shift+e',
    'pane:splitHorizontal': 'cmd+shift+n',
    'pane:splitVertical': 'cmd+n',
    'pane:close': 'cmd+w',
    'editor:undo': 'cmd+z',
    'editor:redo': 'cmd+shift+z',
    'editor:cut': 'cmd+x',
    'editor:copy': 'cmd+c',
    'editor:paste': 'cmd+v',
    'editor:break': 'meta+c',
    'window:close': 'cmd+q',
    'editor:reload': 'cmd+r',
    'pane:next': 'cmd+d',
    'pane:prev': 'cmd+shift+d',

    "editor:break": "cmd+esc",
    "editor:movePreviousWord": "cmd+left",
    "editor:moveNextWord": "cmd+right",
    "editor:moveBeginningLine": "alt+left",
    "editor:moveEndLine": "alt+right",
    "editor:deletePreviousWord": "cmd+backspace",
    "editor:deleteNextWord": "cmd+delete",
    "editor:deleteBeginningLine": "alt+backspace",
    "editor:deleteEndLine": "alt+delete",
  },
};
```

And `~/.zshrc` too fwiw:
```
export ZSH="$HOME/.oh-my-zsh"

ZSH_THEME="powerlevel9k/powerlevel9k"
ZSH_DISABLE_COMPFIX=false

POWERLEVEL9K_MODE='nerdfont-complete'
POWERLEVEL9K_STATUS_VERBOSE=false
POWERLEVEL9K_LEFT_PROMPT_ELEMENTS=(ssh context dir status) # os_icon
POWERLEVEL9K_SHORTEN_STRATEGY='truncate_from_right'
POWERLEVEL9K_SHORTEN_DIR_LENGTH=10
POWERLEVEL9K_CONTEXT_TEMPLATE="`hostname -f | sed "s/.local//" | sed "s/.airgc.net//"`"
POWERLEVEL9K_RIGHT_PROMPT_ELEMENTS=(os_icon vcs root_indicator) # context
POWERLEVEL9K_VCS_GIT_HOOKS=(git-remotebranch)

plugins=(
  git
  sublime
  history
  kubectl
  colorize
  colored-man-pages
  zsh-syntax-highlighting
  zsh-autosuggestions
)

source $ZSH/oh-my-zsh.sh

# FZF
if [[ ! -f ~/.fzf.zsh ]]; then
  echo "Installing fzf..."
  git clone --depth 1 https://github.com/junegunn/fzf.git ~/.fzf
  ~/.fzf/install
fi
source ~/.fzf.zsh
```

**Comment by Stanzilla** [comment]
Can you try just removing/renaming your config and see if it works with the default one?

**Comment by zaneclaes** [comment]
Still no go. 

See `3.0.2` running on the right side (second screenshot), and note the commands I ran in order to remove `~/.hyper.js`. The left window, `3.1.0`, was opened after that was ran. Also, note that the display but does not happen right away. It appears to boot fine, including the prompt (first screenshot). It even renders the nerdfonts correctly, with the Mac icon on the right. The bug happens _as soon as I press a key or tab away from the window._
<img width="1491" alt="Screen Shot 2020-05-04 at 4 46 21 PM" src="https://user-images.githubusercontent.com/2084821/81020773-d31cc180-8e26-11ea-8147-68d470536e28.png">

<img width="3005" alt="Screen Shot 2020-05-04 at 4 43 32 PM" src="https://user-images.githubusercontent.com/2084821/81020780-d617b200-8e26-11ea-89e1-b8d9aa14c810.png">

**Comment by LabhanshAgrawal** [comment]
@zaneclaes instead of the releases page get the latest build from the most recent run in this page https://github.com/zeit/hyper/actions?query=event%3Apush+branch%3Acanary

The build on the releases page still had some issues iirc, which've been fixed in the ci build

**Comment by zaneclaes** [comment]
Okay, I downloaded build #95769080. This version works with the default config file, but did not work with my config file. So I removed everything one at a time and isolated the problem to the nerdfonts. Specifically, the `fontFamily` line in the config causes the bug.

<img width="1634" alt="Screen Shot 2020-05-05 at 8 14 50 AM" src="https://user-images.githubusercontent.com/2084821/81076135-87126100-8ea8-11ea-8b84-df01e6440777.png">

```
module.exports = {
  config: {
    fontFamily: '"Inconsolata Nerd Font", Menlo, "DejaVu Sans Mono", Consolas, "Lucida Console", monospace',
  },
};
```

**Comment by LabhanshAgrawal** [comment]
@zaneclaes Can you also try once with ligatures disabled, set `disableLigatures: true` in config.

**Comment by zaneclaes** [comment]
That fixed it! Thanks @LabhanshAgrawal — fyi the `nano` pagination bug is confirmed fixed with this too (I disabled the `TERM=screen` export).

**Comment by JoeEarly** [comment]
This should make it for 3.1.0 release?

**Comment by LabhanshAgrawal** [comment]
@zaneclaes Can you reply with the link from where you downloaded the font, I want to try and reproduce the issue, so that we might not need to disable ligatures for it.

**Comment by zaneclaes** [comment]
It's part of the [Nerd Fonts Package](https://www.nerdfonts.com/). IIRC I may have [built it from source](https://github.com/buzzkillhardball/nerdfonts), but the first URL appears to have a much more modular user-friendly approach than when I first started using Nerd Fonts years ago.

**Comment by JoeEarly** [comment]
Is this fixed and live?

**Comment by LabhanshAgrawal** [comment]
You can download a [ci build](https://github.com/vercel/hyper/actions/runs/153185087) , it isn't released yet.

**Comment by deanhouseholder** [comment]
I was able to fix this!

I just compiled the source code from scratch. I cloned the canary branch and followed the instructions on the github readme: https://github.com/vercel/hyper

I believe the problem was with the xterm.js npm package and has been fixed there but because Hyper has not been re-compiled and released since May 2019, the fix has not been distributed.

My working build is 3.1.0-canary.4.
